keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,1235 @@
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+- Some CPU Processor vendors have a multi-chiplet architecture consisting of multiple L3 cache groups. Similar to how certain workloads benefit when deployed using the Static CPU Manager Policy option, some workloads see an additional performance uplift when affined to a single L3 cache or distrubted across fewer L3 caches.
+- The existing Static CPU policy assigns CPUs in a numerical order. The user's pod CPU requirements can vary and won't always be a perfect fit to the CPUs tied to the same L3 cache group.
+- With an uncore cache alignment feature enabled in the static policy, the user can depend on the CPU scheduler to optimally assign CPUs with regards to minimizing distribution across L3 caches. 
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- Feature is enabled by a static policy option flag.  
+- It does not change behavior of non-static policy
+- It preserves the behavior of other static options.
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+- Add `UnCoreCacheID` in `CPUInfo` and modify CPUTopology struct to also track uncorecaches in `topology.go`
+  
+  ```go
+  // CPUTopology contains details of node cpu, where :
+  // CPU  - logical CPU, cadvisor - thread
+  // Core - physical CPU, cadvisor - Core
+  // Socket - socket, cadvisor - Socket
+  // NUMA Node - NUMA cell, cadvisor - Node
+  // UnCoreCacheID - Split L3 Cache Topology, cadvisor
+  type CPUTopology struct {
+  NumCPUs        int
+  NumCores       int
+  NumUnCoreCache int
+  NumSockets     int
+  NumNUMANodes   int
+  CPUDetails     CPUDetails
+  }
+  type CPUInfo struct {
+    NUMANodeID int
+    SocketID   int
+    CoreID     int
+    UnCoreCacheID  int
+  }
+  ```
+  
+- Get the UnCoreCacheID for each CPU from cadvisor.
+
+  ```go
+  func Discover(machineInfo *cadvisorapi.MachineInfo) (*CPUTopology, error) {
+  CPUDetails[cpu] = CPUInfo{
+    CoreID:     coreID,
+    SocketID:   core.SocketID,
+    NUMANodeID: node.Id,
+    UnCoreCacheID:  core.UncoreCaches[0].Id
```

Comment:
Where should the logic live for validation of multiple l3 caches?

Does cadvisor handle this case correctly? 

Are all CPUs guaranteed to have an UncoreCache? Or do we need to guard against it?

Reply:
1. Having multiple caches is not a use case I'm aware, can you elaborate what you mean by this? Or anyone can elaborate on this to me that would be helpful
2. We are directly using what cadvisor reports back but I am not totally sure if cadvisor handles this case correctly, assuming you mean it returns the correct uncorcache
3. that's a good question it could be a possibility that can be covered as we go imo
4.

Reply:
I saw UnCoreCaches was an array so I thought that there are multiple ones.

Reply:
let's review what cadvisor does and why. The cadvisor API, which returns a slice, can be confusing perhaps.

Reply:
ok, so here's the cadvisor PR: https://github.com/google/cadvisor/pull/2849/commits/8ffe048f01888036be0e6114a51c32aede63082d

There's little context to be inferred from either the code change or the review.
I'm inclined to believe `UncoreCaches` was made a slice to mimc `Caches`. The parsing logic iterates over detected cache blocks and classifies them as uncore or as something else.

This is unfortunately inconclusive atm about the number of L3 caches :\

Reply:
That's unfortunate

From testing I have only seen it filled with one value per CPU for the slice. Do we still want to investigate this or can this be resolved?

Reply:
There is a section for notes, maybe that's where this belongs

Reply:
For alpha, we can assume that if the slice is not-nil, we can pick the zero-th element and use it. For beta and following we should probably revise the cadvisor API and/or rediscuss this.